src/librustc/hir/map/hir_id_validator.rs

```diff
@@ -1,170 +1,48 @@
-    fn visit_item(&mut self, i: &'hir hir::Item) {
-        let mut inner_visitor = self.new_inner_visitor(self.hir_map);
-        inner_visitor.check(i.hir_id, |this| intravisit::walk_item(this, i));
-    }
-
-    fn visit_trait_item(&mut self, i: &'hir hir::TraitItem) {
-        let mut inner_visitor = self.new_inner_visitor(self.hir_map);
-        inner_visitor.check(i.hir_id, |this| intravisit::walk_trait_item(this, i));
-    }
-
-    fn visit_impl_item(&mut self, i: &'hir hir::ImplItem) {
-        let mut inner_visitor = self.new_inner_visitor(self.hir_map);
-        inner_visitor.check(i.hir_id, |this| intravisit::walk_impl_item(this, i));
-    }
-}
-
-impl<'a, 'hir> HirIdValidator<'a, 'hir> {
-    #[cold]
-    #[inline(never)]
-    fn error(&self, f: impl FnOnce() -> String) {
-        self.errors.lock().push(f());
-    }
-
-    fn check<F: FnOnce(&mut HirIdValidator<'a, 'hir>)>(&mut self,
-                                                       hir_id: HirId,
-                                                       walk: F) {
-        assert!(self.owner_def_index.is_none());
-        let owner_def_index = self.hir_map.local_def_id(hir_id).index;
-        self.owner_def_index = Some(owner_def_index);
-        walk(self);
-
-        if owner_def_index == CRATE_DEF_INDEX {
-            return;
-        }
-
-        // There's always at least one entry for the owning item itself
-        let max = self.hir_ids_seen
-                      .iter()
-                      .map(|local_id| local_id.as_usize())
-                      .max()
-                      .expect("owning item has no entry");
-
-        if max != self.hir_ids_seen.len() - 1 {
-            // Collect the missing ItemLocalIds
-            let missing: Vec<_> = (0 ..= max as u32)
-              .filter(|&i| !self.hir_ids_seen.contains(&ItemLocalId::from_u32(i)))
-              .collect();
-
-            // Try to map those to something more useful
-            let mut missing_items = Vec::with_capacity(missing.len());
-
-            for local_id in missing {
-                let hir_id = HirId {
-                    owner: owner_def_index,
-                    local_id: ItemLocalId::from_u32(local_id),
-                };
-
-                trace!("missing hir id {:#?}", hir_id);
-
-                missing_items.push(format!("[local_id: {}, node:{}]",
-                                           local_id,
-                                           self.hir_map.node_to_string(hir_id)));
+    par_iter(hir_map.map.iter_enumerated()).for_each(|(owner, local_map)| {
+        // Collect the missing `ItemLocalId`s.
+        let missing: Vec<_> = local_map
+            .iter_enumerated()
+            .filter(|(_, entry)| entry.is_none())
+            .map(|(local_id, _)| local_id)
+            .collect();
+
+        if !missing.is_empty() {
+            // HACK(eddyb) avoiding showing more than one error:
+            if !errors.lock().is_empty() {
+                return;
             }
```

Comment:
Oops, this was for avoiding a flood of output from libcore, I need to take it out.
However, it's on the error path so I expect it won't make a difference for perf.